src/evm/memory/evm_allocator.zig

```diff
@@ -1,86 +1,86 @@
-    parent_allocator: std.mem.Allocator,
+    /// The backing allocator - either page allocator or wasm allocator
+    backing_allocator: std.mem.Allocator,
     
-    /// Arena allocator for actual memory management
-    arena: std.heap.ArenaAllocator,
+    /// Current allocated memory buffer (page-aligned)
+    memory: []align(PAGE_SIZE) u8,
     
     /// Current allocated size in bytes
     allocated_size: usize,
     
-    /// Total capacity in bytes
-    capacity: usize,
-    
-    /// Base memory pointer (page-aligned)
-    base_ptr: [*]u8,
-    
     /// Whether to use doubling strategy for growth
     use_doubling_strategy: bool,
 
-    /// Initialize the EVM memory allocator with initial capacity
-    pub fn init(parent_allocator: std.mem.Allocator) !Self {
-        var arena = std.heap.ArenaAllocator.init(parent_allocator);
-        errdefer arena.deinit();
-
+    /// Initialize the EVM memory allocator
+    /// Automatically selects the appropriate backing allocator based on target
+    pub fn init() !Self {
+        // Select appropriate allocator based on target
+        const backing_allocator = if (builtin.target.cpu.arch == .wasm32 or builtin.target.cpu.arch == .wasm64)
+            std.heap.wasm_allocator
+        else
+            std.heap.page_allocator;
+        
         // Allocate initial memory with page alignment
-        const initial_memory = try arena.allocator().alignedAlloc(u8, PAGE_SIZE, INITIAL_SIZE);
+        const initial_memory = try backing_allocator.alignedAlloc(u8, PAGE_SIZE, INITIAL_SIZE);
         
         return Self{
-            .parent_allocator = parent_allocator,
-            .arena = arena,
+            .backing_allocator = backing_allocator,
+            .memory = initial_memory,
             .allocated_size = 0,
-            .capacity = INITIAL_SIZE,
-            .base_ptr = initial_memory.ptr,
             .use_doubling_strategy = true,
         };
     }
 
     /// Deinitialize the allocator and free all memory
     pub fn deinit(self: *Self) void {
-        self.arena.deinit();
+        self.backing_allocator.free(self.memory);
     }
 
     /// Reset the allocator without deallocating memory
     /// This is useful for reusing the allocator between contract executions
     pub fn reset(self: *Self) void {
-        self.allocated_size = 0;
         // Clear the memory to avoid data leakage between contracts
-        @memset(self.base_ptr[0..self.capacity], 0);
+        if (self.allocated_size > 0) {
+            @memset(self.memory[0..self.allocated_size], 0);
+        }
+        self.allocated_size = 0;
     }
 
     /// Get a slice of the currently allocated memory
     pub fn getMemory(self: *const Self) []u8 {
-        return self.base_ptr[0..self.allocated_size];
+        return self.memory[0..self.allocated_size];
     }
 
     /// Get the total capacity
     pub fn getCapacity(self: *const Self) usize {
-        return self.capacity;
+        return self.memory.len;
```

Comment:
gonna have to read arena allocator implementation for reference of how to do these right